### ISSUE TITLE
package for epel7

### ISSUE THREAD

**Issue by ralphbean** [fedora-infra/bodhi#142]
I can build bodhi-server for Fedora 21 now :tada:

.. but epel7 needs a handful more deps to be branched.  Namely:

- [x] ~~pcaro-hermit-fonts~~
- [x] python-colander
- [x] python-cornice [rhbz ticket](https://bugzilla.redhat.com/show_bug.cgi?id=1213968)
- [x] python-cryptography (stuck on this [rhbz ticket](https://bugzilla.redhat.com/show_bug.cgi?id=1185409#c14))
- [x] ~~python-nose-cov~~
- [x] python-simplemediawiki (requested epel7 branch [here](https://bugzilla.redhat.com/show_bug.cgi?id=1213979))
- [x] python-waitress (requested epel7 branch [here](https://bugzilla.redhat.com/show_bug.cgi?id=1213981))

**Comment by lmacken** [comment]
I think we'll need `python-rxjson ` as well for `python-cornice`? We could potentially patch that out as well, but I'm not quite sure yet.

**Comment by ralphbean** [comment]
Looks like rxjson is already branched, but just not built:  https://admin.fedoraproject.org/pkgdb/package/python-rxjson/

**Comment by lmacken** [comment]
* [X] https://admin.fedoraproject.org/updates/python-rxjson-0.2-2.el7

**Comment by lmacken** [comment]
I setup a bodhi2 copr that I used to setup the bodhi02.stg instance

https://copr.fedoraproject.org/coprs/lmacken/bodhi2/

**Comment by ralphbean** [comment]
Nice!

**Comment by lmacken** [comment]
I think we're good to go here. I think it's just a matter of creating the EPEL7 branch and trying to build it.

**Comment by bowlofeggs** [comment]
@lmacken I finally found some time to work on this. There are still three dependency issues:

```
Error: Package: bodhi-server-0.9.8-6.el7.centos.noarch (/bodhi-server-0.9.8-6.el7.centos.noarch)
           Requires: python-tgcaptcha2
Error: Package: bodhi-server-0.9.8-6.el7.centos.noarch (/bodhi-server-0.9.8-6.el7.centos.noarch)
           Requires: python-turboflot
Error: Package: bodhi-server-0.9.8-6.el7.centos.noarch (/bodhi-server-0.9.8-6.el7.centos.noarch)
           Requires: python-fedora-turbogears
```

In full disclosure, I basically just copied a lot of code from the Rawhide spec file and then seasoned them to taste for EL 7. So next I did some grepping to see whether these dependencies are real for the older version of Bodhi in EL 7 or not. Here's what I found for each:

# python-tgcaptcha2

I do see import statements in the code for tgcaptcha, but not tgcaptcha2. However, the spec file included with the code also lists it with the 2, so is it the case the tgcaptcha2 is imported as simply tgcaptcha? I did a test on F24, but it does seem that tgcaptcha and tgcaptcha2 are separate Python packages:

```
[rbarlow@ohm ~]$ dnf provides /usr/lib/python2.7/site-packages/tgcaptcha
Last metadata expiration check: 0:03:23 ago on Tue Jul 26 12:06:33 2016.
python-tgcaptcha-0.11-17.fc24.noarch : TurboGears CAPTCHA widget for forms
Repo        : fedora

[rbarlow@ohm ~]$ dnf provides /usr/lib/python2.7/site-packages/tgcaptcha2
Last metadata expiration check: 0:03:29 ago on Tue Jul 26 12:06:33 2016.
python-tgcaptcha2-0.3.1-1.fc24.noarch : TurboGears captcha plugin
Repo        : fedora
```

Either way, it seems that neither of them is packaged for EL 7.

# python-turboflot

I do see turboflot being imported, so we'll need to get this into EL 7 as well.

# python-fedora-turbogears

EL 7 does have a python-fedora-turbogears2, but not a python-fedora-turbogears. When I grep the code for turbogears references, I don't see any that have "fedora" on the import line. I do see it in the spec file. Is it possible that this dependency is no longer needed?

**Comment by bowlofeggs** [comment]
After talking with @lmacken on Freenode, I learned that the goal is also to update EPEL 7 from the 0.9 series to the new 2.1 series. Thus the concerns I listed above are not relevant. However, this plan comes with a whole set of new concerns!

# EPEL update policy

According to the [EPEL update policy](https://fedoraproject.org/wiki/EPEL_incompatible_upgrades_policy) we may not be allowed to update the bodhi package in EPEL 7 to the latest version since it is not backwards compatible with the old version. However, I [started a thread](https://lists.fedoraproject.org/archives/list/epel-devel@lists.fedoraproject.org/thread/CQFQTIBGRGBDDFSA5PVHTMGNXL5DECHM/) on the epel-devel mailing list to request an exception to that policy. We have to wait at least a week to see what happen on that thread. If they deny the request, we will probably need to add a new bodhi2 package, which is fine.

# Rawhide

Rawhide also has version 0.9 of bodhi, and so it wouldn't make sense to update bodhi on EPEL 7 to be newer than what Fedora has. Due to this, I decided to go ahead and work on getting Rawhide up to date with the latest bodhi release (2.1.8). I decided to use the spec file from @lmacken's [COPR](http://copr-dist-git.fedorainfracloud.org/cgit/lmacken/bodhi2/bodhi.git/tree/bodhi.spec?id=7bd9096bbaa7fb5adfe2af3743603a7ee022e4e2) as a starting point. I made a few tweaks to get it to build successfully in Rawhide and EPEL 7, including adding a patch to make sure the setup.py had the correct version. I also submitted #877 to fix that upstream.

While working on the Rawhide package, I noticed that the man page is no longer being built for the bodhi CLI. Since the new CLI is different than the old CLI, I believe this is a blocker to releasing the new Bodhi CLI in Rawhide. I attempted to issue a `make man` in the docs directory during the build, but ran into a variety of issues. I then noticed that there was a comment that we'd need to use help2man or similar to generate a man page for the new CLI, and so I came to the conclusion that the docs folder probably isn't going to build a correct man page. Thus, before we issue a fedpkg build on Rawhide, we need to get a man page written somehow so we have something to ship.

Additionally, @lmacken noted that we will need to make a 1-line change to fedpkg so that it knows how to work with the new bodhi CLI. To my knowledge, that change hasn't happened in upstream fedpkg yet. We will need to coordinate building/releasing the new bodhi CLI with the fedpkg update that will work with it.

I have pushed up a [commit with my changes](http://pkgs.fedoraproject.org/cgit/rpms/bodhi.git/commit/?id=617d3a96d8d801efc2a4e5e88099b1a235251633) to Fedora's SCM, but have not built it yet since we need to coordinate with fedpkg.

# Fedora 23, Fedora 24 and Fedora 25

Fedora 23, Fedora 24, and Fedora 25 have the bodhi-0.9 package. I debated updating them to the new 2.1 series as well, especially since we are aiming to do so in EPEL 7. However, for now I've come to the conclusion that we should leave 23, 24 and 25 alone, since they currently work "well enough" with the production Bodhi server and the stable releases each have about 6 and 12 months of support left (where EPEL 7 has many years left). Though Fedora 25 is not out yet, we have now passed the alpha branching date and so I wish to avoid altering its bodhi package in any significant way. In other words, let's leave them alone until there is a known issue with not doing so, so as not to introduce non-necessary backwards-incompatible changes.

# Next steps

Our next steps are:
- [x] Write a man page for the new bodhi CLI.
- [x] Submit a pull request against fedpkg so that it works with the new bodhi CLI.
- [x] Add a --version flag to the bodhi 2 CLI: #895.
- [x] Submit a PR against fedpkg that uses the --version flag to work with Bodhi 2 or Bodhi 1, whichever is installed.
- [x] Coordinate a build/release of the new bodhi and fedpkg packages for Rawhide.
- [x] Follow up on the EPEL backwards-incompatible change request thread on August 3.
- [x] Test the new bodhi package in Rawhide and EPEL 7.
- [x] Announce the change to fedora-devel, epel-announce, and epel-devel once it is live in Rawhide and EPEL 7.

**Comment by bowlofeggs** [comment]
Since the Fedora 25 alpha freeze date has now passed, I updated my previous comment to say that we will only release the new bodhi package in Rawhide and EPEL 7.

**Comment by bowlofeggs** [comment]
We were approved to release the backwards-incompatible change during today's EPEL meeting.

**Comment by bowlofeggs** [comment]
5a here is where it was listed that it was agreed for us to update bodhi in EPEL 7:

https://meetbot.fedoraproject.org/fedora-meeting/2016-08-10/epel.2016-08-10-18.00.html

**Comment by bowlofeggs** [comment]
I've submitted a pull request that adds a basic man page for the CLI: https://github.com/fedora-infra/bodhi/pull/883

**Comment by bowlofeggs** [comment]
This merged pull request makes fedpkg compatible with Bodhi 0.9 and 2.1 at the same time, which gives us more flexibility with release timing (i.e., we don't have to "turn our keys" at the same time as fedpkg):

https://pagure.io/fedpkg/pull-request/40

I've now filed BZs against Rawhide and EPEL 7 to request that they update fedpkg:

https://bugzilla.redhat.com/show_bug.cgi?id=1371996
https://bugzilla.redhat.com/show_bug.cgi?id=1371998

**Comment by bowlofeggs** [comment]
https://github.com/fedora-infra/bodhi/issues/900 is also a blocker for this work.

**Comment by bowlofeggs** [comment]
I've deployed a 2.1.9 beta build to https://bodhi.stg.fedoraproject.org/ and would appreciate testing feedback.

**Comment by bowlofeggs** [comment]
2.2.0 is now built in Rawhide.

**Comment by bowlofeggs** [comment]
There is now a Bodhi update for bodhi 2.2.0 in EPEL 7. I'm going to go ahead and close this ticket.

https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2016-b261ba8af3
